File: annotation-processor/src/main/java/com/github/alex/meta/processor/GetterSetterProcessor.java
package com.github.alex.meta.processor;

import com.github.alex.meta.annotation.Getter;
import com.google.auto.service.AutoService;
import com.sun.source.tree.Tree;
import com.sun.tools.javac.api.JavacTrees;
import com.sun.tools.javac.code.Flags;
import com.sun.tools.javac.code.Type;
import com.sun.tools.javac.processing.JavacProcessingEnvironment;
import com.sun.tools.javac.tree.JCTree;
import com.sun.tools.javac.tree.TreeMaker;
import com.sun.tools.javac.tree.TreeTranslator;
import com.sun.tools.javac.util.*;

import javax.annotation.processing.*;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.Element;
import javax.lang.model.element.TypeElement;
import javax.tools.Diagnostic;
import java.util.Set;

@SupportedSourceVersion(SourceVersion.RELEASE_8)
@SupportedAnnotationTypes({"com.github.alex.meta.annotation.Getter", "com.github.alex.meta.annotation.Setter"})
@AutoService(Processor.class)
public class GetterSetterProcessor extends AbstractProcessor {
    private Messager messager; // 编译时期输入日志的
    private JavacTrees javacTrees; // 提供了待处理的抽象语法树
    private TreeMaker treeMaker; // 封装了创建AST节点的一些方法
    private Names names; // 提供了创建标识符的方法

    @Override
    public synchronized void init(ProcessingEnvironment processingEnv) {
        super.init(processingEnv);
        this.messager = processingEnv.getMessager();
        this.javacTrees = JavacTrees.instance(processingEnv);
        Context context = ((JavacProcessingEnvironment) processingEnv).getContext();
        this.treeMaker = TreeMaker.instance(context);
        this.names = Names.instance(context);
    }

    @Override
    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
        Set<? extends Element> elementsAnnotatedWith = roundEnv.getElementsAnnotatedWith(Getter.class);
        elementsAnnotatedWith.forEach(e -> {
            JCTree tree = javacTrees.getTree(e);
            tree.accept(new TreeTranslator() {
                @Override
                public void visitClassDef(JCTree.JCClassDecl jcClassDecl) {
                    List<JCTree.JCVariableDecl> jcVariableDeclList = List.nil();
                    // 在抽象树中找出所有的变量
                    for (JCTree jcTree : jcClassDecl.defs) {
                        if (jcTree.getKind().equals(Tree.Kind.VARIABLE)) {
                            JCTree.JCVariableDecl jcVariableDecl = (JCTree.JCVariableDecl) jcTree;
                            jcVariableDeclList = jcVariableDeclList.append(jcVariableDecl);
                        }
                    }
                    // 对于变量进行生成方法的操作
                    jcVariableDeclList.forEach(jcVariableDecl -> {
                        messager.printMessage(Diagnostic.Kind.NOTE, jcVariableDecl.getName() + " has been processed");
                        jcClassDecl.defs = jcClassDecl.defs
                                .prepend(makeGetterMethodDecl(jcVariableDecl))
                                .append(makeSetterMethodDecl(jcVariableDecl))
                        ;
                    });
                    super.visitClassDef(jcClassDecl);
                }
            });
        });
        return true;
    }


    private JCTree.JCMethodDecl makeGetterMethodDecl(JCTree.JCVariableDecl jcVariableDecl) {
        ListBuffer<JCTree.JCStatement> statements = new ListBuffer<>();
        // 生成return this.<varName>
        JCTree.JCReturn returnStatement = treeMaker.Return(
                treeMaker.Select(treeMaker.Ident(names.fromString("this")), jcVariableDecl.getName())
        );
        statements.append(returnStatement);
        JCTree.JCBlock block = treeMaker.Block(0, statements.toList());
        //返回值类型
        JCTree.JCExpression methodType = treeMaker.Type(jcVariableDecl.getType().type);
        return treeMaker.MethodDef(treeMaker.Modifiers(Flags.PUBLIC),
                addPrefix("get", jcVariableDecl.getName()), methodType, List.nil(), List.nil(), List.nil(), block, null);
    }


    private JCTree.JCMethodDecl makeSetterMethodDecl(JCTree.JCVariableDecl jcVariableDecl) {
        ListBuffer<JCTree.JCStatement> statements = new ListBuffer<>();

        JCTree.JCExpressionStatement aThis = makeAssignment(treeMaker.Select(treeMaker.Ident(names.fromString("this")), jcVariableDecl.getName()), treeMaker.Ident(jcVariableDecl.getName()));
        statements.append(aThis);
        JCTree.JCBlock block = treeMaker.Block(0, statements.toList());

        JCTree.JCVariableDecl param = treeMaker.VarDef(treeMaker.Modifiers(Flags.PARAMETER), jcVariableDecl.getName(), jcVariableDecl.vartype, null);
        List<JCTree.JCVariableDecl> parameters = List.of(param);

        JCTree.JCExpression methodType = treeMaker.Type(new Type.JCVoidType());
        return treeMaker.MethodDef(treeMaker.Modifiers(Flags.PUBLIC),
                addPrefix("set", jcVariableDecl.getName()), methodType, List.nil(), parameters, List.nil(), block, null);
    }

    private Name addPrefix(String prefix, Name name) {
        String s = name.toString();
        return names.fromString(prefix + s.substring(0, 1).toUpperCase() + s.substring(1, name.length()));
    }

    private JCTree.JCExpressionStatement makeAssignment(JCTree.JCExpression lhs, JCTree.JCExpression rhs) {
        return treeMaker.Exec(treeMaker.Assign(lhs, rhs));
    }
}

File: kapt-annotation-processor/src/main/kotlin/com/github/alex/kapt/meta/processor/GetterSetterProcessor.kt
@file:Suppress("JAVA_MODULE_DOES_NOT_EXPORT_PACKAGE")
package com.github.alex.kapt.meta.processor

import com.github.alex.kapt.meta.annotation.Getter
import com.sun.source.tree.Tree
import com.sun.tools.javac.api.JavacTrees
import com.sun.tools.javac.code.Flags
import com.sun.tools.javac.code.Type.JCVoidType
import com.sun.tools.javac.processing.JavacProcessingEnvironment
import com.sun.tools.javac.tree.JCTree.*
import com.sun.tools.javac.tree.TreeMaker
import com.sun.tools.javac.tree.TreeTranslator
import com.sun.tools.javac.util.List
import com.sun.tools.javac.util.ListBuffer
import com.sun.tools.javac.util.Name
import com.sun.tools.javac.util.Names
import java.lang.RuntimeException
import java.util.*
import javax.annotation.processing.*
import javax.annotation.processing.Processor
import javax.lang.model.SourceVersion
import javax.lang.model.element.Element
import javax.lang.model.element.TypeElement
import javax.tools.Diagnostic

@SupportedSourceVersion(SourceVersion.RELEASE_11)
@SupportedAnnotationTypes("com.github.alex.kapt.meta.annotation.Getter", "com.github.alex.kapt.meta.annotation.Setter")
@com.google.auto.service.AutoService(Processor::class)
class GetterSetterProcessor: AbstractProcessor() {
    private var messager: Messager? = null // 编译时期输入日志的

    private var javacTrees: JavacTrees? = null // 提供了待处理的抽象语法树

    private var treeMaker: TreeMaker? = null // 封装了创建AST节点的一些方法

    private var names: Names? = null // 提供了创建标识符的方法


    override fun init(processingEnv: ProcessingEnvironment) {
        super.init(processingEnv)
        messager = processingEnv.messager
        javacTrees = JavacTrees.instance(processingEnv)
        val context = (processingEnv as JavacProcessingEnvironment).context
        treeMaker = TreeMaker.instance(context)
        names = Names.instance(context)
    }
    override fun process(annotations: Set<TypeElement?>?, roundEnv: RoundEnvironment): Boolean {
        val elementsAnnotatedWith: Set<Element?> = roundEnv.getElementsAnnotatedWith(Getter::class.java)
        elementsAnnotatedWith.forEach { e: Element? ->
            val tree = javacTrees!!.getTree(e)
            tree.accept(object : TreeTranslator() {
                override fun visitClassDef(jcClassDecl: JCClassDecl) {
                    var jcVariableDeclList =
                        List.nil<JCVariableDecl>()
                    // 在抽象树中找出所有的变量
                    for (jcTree in jcClassDecl.defs) {
                        if (jcTree.kind == Tree.Kind.VARIABLE) {
                            val jcVariableDecl = jcTree as JCVariableDecl
                            jcVariableDeclList = jcVariableDeclList.append(jcVariableDecl)
                        }
                    }

                    // 对于变量进行生成方法的操作
                    jcVariableDeclList.forEach { jcVariableDecl: JCVariableDecl ->
                        messager!!.printMessage(
                            Diagnostic.Kind.NOTE,
                            jcVariableDecl.getName().toString() + " has been processed"
                        )
                        jcClassDecl.defs = jcClassDecl.defs
                            .prepend(makeGetterMethodDecl(jcVariableDecl))
                            .append(makeSetterMethodDecl(jcVariableDecl))
                    }
                    super.visitClassDef(jcClassDecl)
                }
            })
        }
        return true
    }


    private fun makeGetterMethodDecl(jcVariableDecl: JCVariableDecl): JCMethodDecl {
        val statements = ListBuffer<JCStatement>()
        // 生成return this.<varName>
        val returnStatement = treeMaker!!.Return(
            treeMaker!!.Select(treeMaker!!.Ident(names!!._this), jcVariableDecl.getName())
        )
        statements.append(returnStatement)
        val block = treeMaker!!.Block(0, statements.toList())
        //返回值类型
        val methodType = treeMaker!!.Type(jcVariableDecl.getType().type)
        return treeMaker!!.MethodDef(
            treeMaker!!.Modifiers(Flags.PUBLIC.toLong()),
            addPrefix("get", jcVariableDecl.getName()), methodType, List.nil(), List.nil(), List.nil(), block, null
        )
    }


    private fun makeSetterMethodDecl(jcVariableDecl: JCVariableDecl): JCMethodDecl {
        val statements = ListBuffer<JCStatement>()
        val aThis = makeAssignment(
            treeMaker!!.Select(treeMaker!!.Ident(names!!.fromString("this")), jcVariableDecl.getName()),
            treeMaker!!.Ident(jcVariableDecl.getName())
        )
        statements.append(aThis)
        val block = treeMaker!!.Block(0, statements.toList())
        val param = treeMaker!!.VarDef(
            treeMaker!!.Modifiers(Flags.PARAMETER),
            jcVariableDecl.getName(),
            jcVariableDecl.vartype,
            null
        )
        val parameters = List.of(param)
        val methodType = treeMaker!!.Type(JCVoidType())
        return treeMaker!!.MethodDef(
            treeMaker!!.Modifiers(Flags.PUBLIC.toLong()),
            addPrefix("set", jcVariableDecl.getName()), methodType, List.nil(), parameters, List.nil(), block, null
        )
    }

    private fun addPrefix(prefix: String, name: Name): Name {
        val s = name.toString()
        return names!!.fromString(
            prefix + s.substring(0, 1).uppercase(Locale.getDefault()) + s.substring(
                1,
                name.length
            )
        )
    }

    private fun makeAssignment(lhs: JCExpression, rhs: JCExpression): JCExpressionStatement {
        return treeMaker!!.Exec(treeMaker!!.Assign(lhs, rhs))
    }

}